cdc/puller/sorter/heap_sorter.go

```diff
@@ -121,17 +144,16 @@ func (h *heapSorter) flush(ctx context.Context, maxResolvedTs uint64) error {
 		maxResolvedTs: maxResolvedTs,
 		finished:      finishCh,
 		canceller:     h.canceller,
+		isEmpty:       isEmptyFlush,
 	}
 	h.taskCounter++
 
 	var oldHeap sortHeap
 	if !isEmptyFlush {
 		task.dealloc = func() error {
-			if atomic.SwapInt32(&task.isDeallocated, 1) == 1 {
-				return nil
-			}
-			if task.backend != nil {
-				task.backend = nil
+			backEnd := task.GetBackEnd()
+			if backEnd != nil {
+				defer task.markDeallocated()
```

Comment:
What happens if
1. In L154-L155, `backEnd` is not nil
2. When calling `pool.dealloc(backEnd)`, task is marked as deallocated and `task.backend` becomes nil, through `backEnd` is not nil

Reply:
`dealloc` will not race with itself, because `dealloc` calls from different goroutines are always synchronized by a write to `finished` channel.